Log tracing

swalchemist opened the pull request:
Log tracing - this changes the UAA log format used in a bosh uaa deployment to include a trace ID and span ID, and adjusts the tests to match.

This should be merged after https://github.com/cloudfoundry/uaa/pull/2446 in order for the tests to pass. See that PR for details about the feature.

[#185000950]

swalchemist committed:
feat: add trace info to uaa logs
    * The traceId and spanId are populated by the Brave library.

Co-authored-by: Bruce Ricard <bricard@vmware.com>

swalchemist committed:
fix: adapt regex for log tracing format

* Necessary after the test was fixed so that the regex no longer slurps both [] pairs.

bruce-ricard commented:
Do we want to have a separate UAA repo PR for the source code changes, rather than reviewing them from here through the submodule? I find it a bit awkward. 

I was the one who bumped the submodule in the uaa-release feature branch, though it was for ease of developer use while working on it, not necessarily for the PR.

edit: I think that we should make a separate commit: it's essentially impossible to comment on the UAA changes.

bruce-ricard commented on jobs/uaa/templates/config/log4j2.properties.erb:
```diff
@@ -15,7 +15,7 @@ case logging_format_timestamp_value
 %>
 property.log_directory = /var/vcap/sys/log/uaa
-property.log_pattern=[<%= timestamp_format %>] uaa%X{context} - %pid [%t] .... %5p --- %c{1}: %replace{%m}{(?<=password=|client_secret=)([^&]*)}{<redacted>}%n
+property.log_pattern=[<%= timestamp_format %>] uaa%X{context} - %pid [%t] - [%X{traceId},%X{spanId}] .... %5p --- %c{1}: %replace{%m}{(?<=password=|client_secret=)([^&]*)}{<redacted>}%n
```
This is technically a breaking change. Do we want to make it configurable in the spec? @strehle any thoughts?

strehle replied:
in what sense breaking ? but if you have doubts then yes please add a config for it

bruce-ricard replied:
It could possibly break a user's log parser. Although it would be pretty easy to fix.

We don't really want to make it configurable, we think it's OK.

swalchemist replied:
Anything that changes the log format could break existing code that parses the logs. We asked around internally and no one was concerned about it as long as we mentioned the change in the release notes - it seems that our customers are used to updating their log parsing when they install an update. 

So I'm inclined not to make this configurable unless someone has a solid need for it.

Tallicia replied:
It appears this is resolved at this point?

swalchemist commented:
> Do we want to have a separate UAA repo PR for the source code changes, rather than reviewing them from here through the submodule? I find it a bit awkward.

There is a separate PR for UAA - https://github.com/cloudfoundry/uaa/pull/2446. Committing the submodule change here was probably not helpful, and we shouldn't review specific uaa changes in a uaa-release PR anyway.

bruce-ricard commented:
I removed the UAA change from this PR. We now need to merge the UAA one before we merge this one.

hsinn0 commented on scripts/create-dev-release.sh:
```diff
@@ -3,6 +3,9 @@
 set -euo pipefail
 
 export UAA_VERSION=0.0.0
+JAVA_HOME=$(/usr/libexec/java_home -v 11)
```
Why this change? The script should not be tied with specific Java version.

hsinn0 replied:
Also, this is likely to work only on MacOS. It should be reverted.

Tallicia replied:
@swalchemist Can you remove this?

Tallicia reviewed (commented):
These commits look extraneous to the PR feature request.

swalchemist commented:
I have removed a giant mess of unnecessary commits from this branch. I'm not aware of any actual code change that this causes, but I'm not sure that nothing changed.

swalchemist committed:
fix: comparison data for logging tests

Odd that the log4j pattern must be repeated in so many places.

swalchemist committed:
fix: restore import

Trying to fix test failure in CI: "./uaa-release_test.go:48:3: undefined: Expect".

hsinn0 commented on src/acceptance_tests/uaa-release_test.go:
```diff
@@ -140,9 +140,12 @@ var _ = Describe("UaaRelease", func() {
+	DescribeTable("UAA log format", func(
+		uaaLogFormat string,
+		optFiles ...string,
+	) {
```
I'd rather not see this kind of reformatting as part of the PR.

swalchemist replied:
Hmmm, interesting. For Go, the reformatting is usually automatic and something that's very difficult to suppress. So the code this was changed probably wasn't changed with an IDE.

Tallicia replied:
If it's coming from the IDE, and they aren't intended line changes, they can can be unselected.

hsinn0 commented on src/acceptance_tests/uaa-release_test.go:
```diff
@@ -292,7 +295,10 @@ func scpUaaAuditLog() string {
+func getFingerPrint(certdata []byte) (
+	string,
+	error,
+) {
```
The same. I'd rather not see this kind of reformatting as part of the PR.
